src/fsharp/service/IncrementalBuild.fs

```diff
@@ -1506,188 +583,298 @@ type IncrementalBuilder(tcGlobals, frameworkTcImports, nonFrameworkAssemblyInput
+                    typeCheckCache.[slot] <- Checked(sourceFile, tcAcc, WeakReference<_> parsedInfo)
+                    return tcAcc
+                | _ ->
+                    if ct.IsCancellationRequested then
+                        return! Cancellable.canceled ()
+                    else
+                        return invalidOp "Type checking was not canceled." }
+        | Checked(sourceFile, tcAcc, weakInfo) ->
+            match op with
+            | UseCache -> Cancellable.ret tcAcc
+            | ReTypeCheck ->
+                match weakInfo.TryGetTarget() with
+                | true, parsedInfo ->
+                    match parsedInfo.Input with
+                    | Some _ ->
+                        typeCheckCache.[slot] <- Parsed (sourceFile, parsedInfo)
+                    | _ ->
+                        typeCheckCache.[slot] <- NotParsed sourceFile
+                | _ ->
+                    typeCheckCache.[slot] <- NotParsed sourceFile
+
+                typeCheck ctok UseCache slot
+
+    and priorTypeCheck ctok op slot =
+        if IncrementalBuild.injectCancellationFault then Cancellable.canceled ()
+        else
+
+        match slot with
+        | 0 (* first file *) -> getInitialTcAcc ctok
+        | _ -> 
+            cancellable {
+                let slot = slot - 1
+                checkSlot slot
+                match typeCheckCache.[slot] with
+                | Checked (_, tcAcc, _) -> return tcAcc
+                | _ ->
+                    // Find the first slot that has not been checked.
+                    let startingSlot = typeCheckCache |> Array.findIndex (function Checked _ -> false | _ -> true)
+                    let! tcAccs = cancellable { for i = startingSlot to slot do yield! typeCheck ctok op i }
+                    if tcAccs.IsEmpty then
+                        failwith "Should not happen. Expected at least one type-checked result."
+                    return List.last tcAccs }
+
+    let step ctok =
+        cancellable {
+            match typeCheckCache |> Array.tryFindIndex (function NotParsed _ | Parsed _ -> true | _ -> false) with
+            | Some nextSlotToTypeCheck ->
+                let! result = typeCheck ctok UseCache nextSlotToTypeCheck
+                return Some result
+            | _ ->
+                return None }
+
+    let isChecked slot =
+        checkSlot slot
+
+        match typeCheckCache.[slot] with
+        | Checked _ -> true
+        | _ -> false
+
+    let invalidateSlot slot =
+        checkSlot slot
+
+        for i = slot to typeCheckCache.Length - 1 do
+            typeCheckCache.[i] <- NotParsed typeCheckCache.[i].SourceFile
+
+    let invalidateBuild () =
+        initialTcAccCache <- None
+        invalidateSlot 0
+
+    let getSlot fileName =
+        typeCheckCache 
+        |> Array.findIndex (fun x ->
+            let f2 = x.FileName
+            String.Compare(fileName, f2, StringComparison.CurrentCultureIgnoreCase)=0
```

Comment:
There's also an `Equals` overload with a `StringComparison` input that might be a bit more clear here

Reply:
Yea, I agree. This was copied from the original but can be simplified further.